Python/ceval.c

```diff
@@ -3763,10 +3769,14 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
         /* Log traceback info. */
         PyTraceBack_Here(f);
 
-        if (tstate->c_tracefunc != NULL)
+        if (tstate->c_tracefunc != NULL) {
+            /* Temporarily set frame state to RAISED for benefit of frame.setlineno */
+            assert(f->f_state == FRAME_EXECUTING);
+            f->f_state = FRAME_RAISED;
```

Comment:
It seems I missed another backwards state transition.  Maybe it would be good to explain the state machine where the states names are defined.

Reply:
This is a bit of a hack, to keep `frame.setlineno` happy.
I've added `FRAME_UNWINDING` to mark that an exception is being unwound in the frame, but that the frame has not completed yet.